File: app/src/main/java/fr/xgouchet/chronorg/ui/viewholders/PortalViewHolder.java
package fr.xgouchet.chronorg.ui.viewholders;

import android.support.annotation.NonNull;
import android.view.View;

import org.joda.time.ReadablePeriod;

import fr.xgouchet.chronorg.data.formatters.Formatter;
import fr.xgouchet.chronorg.data.models.Portal;

/**
 * @author Xavier Gouchet
 */
public class PortalViewHolder extends BaseProjectItemViewHolder<Portal> {

    @NonNull final Formatter<ReadablePeriod> formatter;

    public PortalViewHolder(@NonNull Listener<Portal> listener,
                            View itemView,
                            @NonNull Formatter<ReadablePeriod> formatter) {
        super(listener, itemView);
        this.formatter = formatter;
    }


    @Override protected String getName(@NonNull Portal portal) {
        return portal.getName();
    }

    @Override protected int getColor(@NonNull Portal portal) {
        return portal.getColor();
    }

    @Override protected String getInfo(@NonNull Portal portal) {
        String dir = "";
        switch (portal.getDirection()) {
            case Portal.Direction.BOTH:
                dir = " ↔";
                break;
            case Portal.Direction.FUTURE:
                dir = " →";
                break;
            case Portal.Direction.PAST:
                dir = " ←";
                break;
        }
        return formatter.format(portal.getDelay().toPeriod()) + dir;
    }
}

File: khronorg/src/main/java/fr/xgouchet/khronorg/feature/portals/PortalViewHolder.kt
package fr.xgouchet.khronorg.feature.portals

import android.view.View
import android.widget.TextView
import fr.xgouchet.khronorg.R
import fr.xgouchet.khronorg.commons.formatters.Formatter
import fr.xgouchet.khronorg.data.models.Direction
import fr.xgouchet.khronorg.ui.Cutelry.knife
import fr.xgouchet.khronorg.ui.viewholders.BaseViewHolder
import io.reactivex.functions.Consumer
import org.joda.time.ReadableDuration
import org.joda.time.ReadableInterval

/**
 * @author Xavier F. Gouchet
 */
class PortalViewHolder(val formatter: Formatter<ReadableInterval>,

                       listener: Consumer<Portal>?, itemView: View)
    : BaseViewHolder<Portal>(listener, itemView) {


    internal val name: TextView by knife(R.id.name, itemView)
    internal val info: TextView by knife(R.id.info, itemView)
    internal val underline: View by knife(R.id.underline, itemView)

    init {
        if (listener != null) {
            itemView.setOnClickListener({ view -> listener.accept(item) })
        }
    }

    override fun onBindItem(item: Portal) {
        name.text = item.name

        val dir: String
        when (item.direction) {
            Direction.FUTURE -> dir = "→"
            Direction.PAST -> dir = "←"
            else -> dir = "?"
        }

        info.text = "${formatter.format(item.delay)} $dir"

        underline.setBackgroundColor(item.color)
    }
}